### PR TITLE
More performance regressions

### PR DESCRIPTION
**Pull request by radumereuta**
These are a couple more commits that I would like to revert since it made the C-semantics take significantly longer times to kompile. It doubled the "Apply compile pipeline" step.
Here are some times with before and after when compiling the C semantics:

f06da998f35d3a7a6b93cf6771ccacccf285a596 master
```
kompile haskell c-semantics ...
Outer parsing [237 modules]                                  =  2.100s
Parse configurations [3/3 declarations]                      =  4.340s
Parse rules [2622/2622 rules]                                = 23.778s
Validate definition                                          =  0.863s
Apply compile pipeline                                       = 40.514s
Kompile to kore                                              =  2.698s
Save to disk                                                 =  1.609s
  Print definition.kore                                      = 14.996s
  Validate def                                               = 25.579s
Backend                                                      = 40.575s
Cleanup                                                      =  0.000s
Total                                                        = 01m 56s
kompile llvm c-semantics ...
Outer parsing [237 modules]                                  =  2.044s
Parse configurations [3/3 declarations]                      =  2.264s
Parse rules [3064/3064 rules]                                = 27.763s
Validate definition                                          =  0.946s
Apply compile pipeline                                       = 21.873s
Kompile to kore                                              =  2.697s
Save to disk                                                 =  1.880s
  Print definition.kore                                      = 16.479s
  Write decision tree                                        = 11.900s
  └interpreter: main                                         = 24.308s
Backend                                                      = 52.688s
Cleanup                                                      =  0.000s
Total                                                        = 01m 52s
Combined time: 3m49,462s
```
7fe67f1e6f8e5b6cf6d220ae04e15c05c7801ae8 - before the changes
```
kompile haskell c-semantics ...
Outer parsing [237 modules]                                  =  0.883s
Parse configurations [3/3 declarations]                      =  1.646s
Parse rules [2622/2622 rules]                                = 24.542s
Validate definition                                          =  0.893s
Apply compile pipeline                                       = 10.365s
Kompile to kore                                              =  2.665s
Save to disk                                                 =  1.650s
  Print definition.kore                                      = 19.256s
  Validate def                                               = 24.628s
Backend                                                      = 43.884s
Cleanup                                                      =  0.000s
Total                                                        = 01m 26s
kompile llvm c-semantics ...
Outer parsing [237 modules]                                  =  0.828s
Parse configurations [3/3 declarations]                      =  1.651s
Parse rules [3064/3064 rules]                                = 27.490s
Validate definition                                          =  0.983s
Apply compile pipeline                                       = 11.004s
Kompile to kore                                              =  2.655s
Save to disk                                                 =  1.741s
  Print definition.kore                                      = 15.953s
  Write decision tree                                        = 11.521s
  └interpreter: main                                         = 32.433s
Backend                                                      = 59.908s
Cleanup                                                      =  0.000s
Total                                                        = 01m 46s
Combined time: 3m15,124s
```
cf035cf644ab58afe1a79d07d431875c5a3e746d - after revert
```
kompile haskell c-semantics ...
Outer parsing [237 modules]                                  =  0.835s
Parse configurations [3/3 declarations]                      =  1.542s
Parse rules [2622/2622 rules]                                = 23.099s
Validate definition                                          =  0.920s
Apply compile pipeline                                       = 10.332s
Kompile to kore                                              =  2.659s
Save to disk                                                 =  1.714s
  Print definition.kore                                      = 14.855s
  Validate def                                               = 25.462s
Backend                                                      = 40.317s
Cleanup                                                      =  0.001s
Total                                                        = 01m 21s
kompile llvm c-semantics ...
Parse command line options                                   =  0.015s
Outer parsing [237 modules]                                  =  0.835s
Parse configurations [3/3 declarations]                      =  1.611s
Parse rules [3064/3064 rules]                                = 26.940s
Validate definition                                          =  0.958s
Apply compile pipeline                                       = 10.832s
Kompile to kore                                              =  2.648s
Save to disk                                                 =  1.896s
  Print definition.kore                                      = 15.623s
  Write decision tree                                        = 11.552s
  └interpreter: main                                         = 24.840s
Backend                                                      = 52.016s
Cleanup                                                      =  0.000s
Total                                                        = 01m 37s
Combined time: 3m1,560s
```

TL;DR these changes doubled the time of applying the compile pipeline and outer parsing (creating the initial module objects).

We should revisit these changes when we have a better understanding on where time is being lost.
cc: @Robertorosmaninho

### CODE REVIEW

**Review by Baltoli — APPROVED** [review]
LGTM as these are just clean reverts. Make sure to open an issue for investigation later as well!

**Comment by radumereuta** [comment]
> LGTM as these are just clean reverts. Make sure to open an issue for investigation later as well!

Fair point. Added a comment in the issue that keeps track of these changes: https://github.com/runtimeverification/k/issues/3444#issuecomment-1665461900
